### PR TITLE
Reduce dependence on commons-lang3

### PR DESCRIPTION
**Pull request by triceo**

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/buildin/bendablebigdecimal/BendableBigDecimalScore.java [hunk 1]
```diff
@@ -402,20 +403,15 @@ public boolean equals(Object o) {
...
-        }
-        for (BigDecimal softScore : softScores) {
-            hashCode = (37 * hashCode) + softScore.stripTrailingZeros().hashCode();
-        }
-        return hashCode;
+        int[] scoreHashCodes = Stream.concat(Arrays.stream(hardScores), Arrays.stream(softScores))
+                .map(BigDecimal::stripTrailingZeros)
+                .mapToInt(BigDecimal::hashCode)
+                .toArray();
+        return Objects.hash(initScore, Arrays.hashCode(scoreHashCodes));
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/buildin/bendablebigdecimal/BendableBigDecimalScore.java]: In my opinion, this is too much to do in `hashCode()`.
In this particular case, I would argue the trailing zeros could be stripped as they come into the `Score`. Precision loss doesn't matter here, as the end value is always going to be the same.

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/buildin/bendablebigdecimal/BendableBigDecimalScore.java]: The stripTrailingZeros is a requirement of the equals() method, which leaks into the hashcode method due to the Java equal contract.
IIRC, the equals method required that due to this case:
BendableScore.ZERO (= "0") must equal BendableScore.ZERO.add("1.00").subtract("1.00") (= "0.00") or it causes score corruption.

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/buildin/bendablebigdecimal/BendableBigDecimalScore.java]: That being said, that stripTrailingZeros does look weird, so the equals() method should really comment that example as a motivation so it doesn't get removed by accident.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/buildin/bendablebigdecimal/BendableBigDecimalScore.java [hunk 2]
```diff
@@ -402,20 +403,15 @@ public boolean equals(Object o) {
...
     }
 
     @Override
     public int compareTo(BendableBigDecimalScore other) {
-        // A direct implementation (instead of CompareToBuilder) to avoid dependencies
         validateCompatible(other);
         if (initScore != other.getInitScore()) {
             return initScore < other.getInitScore() ? -1 : 1;
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/buildin/bendablebigdecimal/BendableBigDecimalScore.java]: (Ok to leave to another PR): this can use Integer.compare(int, int)

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/anchor/AnchorVariableDemand.java [hunk 3]
```diff
@@ -68,7 +68,7 @@ public boolean equals(Object o) {
 
     @Override
     public int hashCode() {
-        return CLASS_NAME_HASH_CODE + sourceVariableDescriptor.hashCode();
+        return Objects.hash(this.getClass().getName(), sourceVariableDescriptor);
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/anchor/AnchorVariableDemand.java]: Change breaks equals contract, because equality doesn't take into account subclass names.
To avoid that, use Objects.hash(AnchorVariableDemand.class.getName(), ...)

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/custom/CustomShadowVariableDemand.java [hunk 4]
```diff
@@ -61,7 +61,7 @@ public boolean equals(Object o) {
 
     @Override
     public int hashCode() {
-        return CLASS_NAME_HASH_CODE + targetShadowVariableDescriptor.hashCode();
+        return Objects.hash(this.getClass().getName(), targetShadowVariableDescriptor);
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/custom/CustomShadowVariableDemand.java]: same

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/inverserelation/CollectionInverseVariableDemand.java [hunk 5]
```diff
@@ -67,7 +67,7 @@ public boolean equals(Object o) {
 
     @Override
     public int hashCode() {
-        return CLASS_NAME_HASH_CODE + sourceVariableDescriptor.hashCode();
+        return Objects.hash(this.getClass().getName(), sourceVariableDescriptor);
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/inverserelation/CollectionInverseVariableDemand.java]: same

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/inverserelation/SingletonInverseVariableDemand.java [hunk 6]
```diff
@@ -64,7 +64,7 @@ public boolean equals(Object o) {
 
     @Override
     public int hashCode() {
-        return CLASS_NAME_HASH_CODE + sourceVariableDescriptor.hashCode();
+        return Objects.hash(this.getClass().getName(), sourceVariableDescriptor);
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/inverserelation/SingletonInverseVariableDemand.java]: same

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/listener/support/VariableListenerNotification.java [hunk 7]
```diff
@@ -58,8 +59,7 @@ public boolean equals(Object o) {
 
     @Override
     public int hashCode() {
-        // A direct implementation (instead of HashCodeBuilder) to avoid dependencies
-        return (((17 * 37) + System.identityHashCode(entity))) * 37 + type.hashCode();
+        return Objects.hash(entity, type);
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/domain/variable/listener/support/VariableListenerNotification.java]: This is not a pure refactor: now the entity's hashcode is used instead of the entity's identityHashCode.
*Breaks equals contract* (and this one unlike the ones above this one will cause trouble): the equals method uses entity same semantics, but the hashcode uses entity equals semantics.
For the record, it needs to remain entity same semantics (the more we force our users to implement equals/hashcode the way we want it, the less compatible we are with other Java frameworks)
PS: we probably want a unit test proving that this breaks the equals contract before we fix it - clearly the original code is missing test coverage :)

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/ChangeMove.java [hunk 8]
```diff
@@ -105,28 +103,22 @@ public String getSimpleMoveTypeDescription() {
...
 
     @Override
-    public boolean equals(Object o) {
+    public boolean equals(final Object o) {
         if (this == o) {
             return true;
-        } else if (o instanceof ChangeMove) {
-            ChangeMove<?> other = (ChangeMove) o;
-            return new EqualsBuilder()
-                    .append(entity, other.entity)
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/ChangeMove.java]: Yea, yea, I know...
This already uses equals() instead of same semantics on entities, but in practice Move equality is only used for Move Tabu Search and nobody uses that. So this hasn't posed anywhere the trouble that VariableListenerNotification would have if you make that equals() semantics.
That being said, move equality should probably use same semantics. Can you create a separate JIRA for that? It will need to be thoroughly tested.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/PillarChangeMove.java [hunk 9]
```diff
@@ -125,28 +123,22 @@ public String getSimpleMoveTypeDescription() {
     }
 
     @Override
-    public boolean equals(Object o) {
+    public boolean equals(final Object o) {
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/PillarChangeMove.java]: (suggestion, non-blocking) Code style: normally we don't put "final" keywords in method parameters.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/chained/KOptMove.java [hunk 10]
```diff
@@ -172,26 +172,23 @@ public String getSimpleMoveTypeDescription() {
...
     }
 
     @Override
     public int hashCode() {
-        return new HashCodeBuilder()
-                .append(entity)
-                .append(values)
-                .toHashCode();
+        int result = Objects.hash(entity);
+        result = 31 * result + Arrays.hashCode(values);
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/chained/KOptMove.java]: Any reason why this doesn't use Objects.hash()?

I am ok for changing the hashcode here as long as it desn't violate the equals contract.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/PillarChangeMove.java [hunk 11]
```diff
@@ -125,28 +123,22 @@ public String getSimpleMoveTypeDescription() {
...
-        } else if (o instanceof PillarChangeMove) {
-            PillarChangeMove<?> other = (PillarChangeMove) o;
-            return new EqualsBuilder()
-                    .append(variableDescriptor, other.variableDescriptor)
-                    .append(pillar, other.pillar)
-                    .append(toPlanningValue, other.toPlanningValue)
-                    .isEquals();
-        } else {
+        }
+        if (o == null || getClass() != o.getClass()) {
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/heuristic/selector/move/generic/PillarChangeMove.java]: +1 to doing the getClass check instead of instanceof. Especially with ChainedChangeMove and ChangeMove it might matter some day.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/uni/BavetFilterUniNode.java [hunk 12]
```diff
@@ -48,9 +49,7 @@ public void addChildNode(BavetAbstractUniNode<A> childNode) {
 
     @Override
     public int hashCode() {
-        // Similar to Object.hash() without autoboxing
-        return 31 * System.identityHashCode(parentNode)
-                + System.identityHashCode(predicate);
+        return Objects.hash(parentNode, predicate);
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/uni/BavetFilterUniNode.java]: IMPORTANT (see above), break equals contract + changes implementation

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/tri/BavetFilterTriNode.java [hunk 13]
```diff
@@ -48,9 +49,7 @@ public void addChildNode(BavetAbstractTriNode<A, B, C> childNode) {
 
     @Override
     public int hashCode() {
-        // Similar to Object.hash() without autoboxing
-        return 31 * System.identityHashCode(parentNode)
-                + System.identityHashCode(predicate);
+        return Objects.hash(parentNode, predicate);
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/tri/BavetFilterTriNode.java]: IMPORTANT (see above), break equals contract + changes implementation

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/tri/BavetFilterTriNode.java]: There a bunch more of these, I won't point them all out.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetFilterBiNode.java [hunk 14]
```diff
@@ -48,9 +49,7 @@ public void addChildNode(BavetAbstractBiNode<A, B> childNode) {
 
     @Override
     public int hashCode() {
-        // Similar to Object.hash() without autoboxing
-        return 31 * System.identityHashCode(parentNode)
-                + System.identityHashCode(predicate);
+        return Objects.hash(parentNode, predicate);
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/bavet/bi/BavetFilterBiNode.java]: IMPORTANT (see above), break equals contract + changes implementation

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/buildin/bendablebigdecimal/BendableBigDecimalScore.java (hunk 1 above)
Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/api/score/buildin/bendablebigdecimal/BendableBigDecimalScore.java]: @ge0ffrey My point is... if we strip all trailing zeros always, we don't need to worry about it in `equals()`.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/cheaptime/solver/drools/MachinePeriodPart.java [hunk 15]
```diff
@@ -97,40 +104,30 @@ public int getPeriod() {
...
-                    .isEquals();
-        } else {
+        }
+        if (o == null || getClass() != o.getClass()) {
             return false;
         }
+        final MachinePeriodPart that = (MachinePeriodPart) o;
+        return period == that.period &&
+                active == that.active &&
+                Objects.equals(machine, that.machine) &&
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/cheaptime/solver/drools/MachinePeriodPart.java]: Code style: please keep original order (machine, period, active), which is the order of the field declarations.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/cheaptime/solver/drools/MachinePeriodPart.java [hunk 16]
```diff
@@ -97,40 +104,30 @@ public int getPeriod() {
...
-                    .append(period, other.period)
-                    .append(active, other.active)
-                    .append(resourceAvailableList, other.resourceAvailableList)
-                    .isEquals();
-        } else {
+        }
+        if (o == null || getClass() != o.getClass()) {
             return false;
         }
+        final MachinePeriodPart that = (MachinePeriodPart) o;
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/cheaptime/solver/drools/MachinePeriodPart.java]: Code nitpick: This renames the variable "other" to "that".
All other equals() code in optaplanner uses "o" for Object and "other" for casted instance.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/travelingtournament/domain/TravelingTournament.java [hunk 17]
```diff
@@ -114,12 +114,8 @@ public boolean equals(Object o) {
 
     @Override
     public int hashCode() {
-        HashCodeBuilder hashCodeBuilder = new HashCodeBuilder();
-        for (Match match : matchList) {
-            // Notice: we don't use hashCode()
-            hashCodeBuilder.append(match.solutionHashCode());
-        }
-        return hashCodeBuilder.toHashCode();
+        int[] hashes = matchList.stream().mapToInt(Match::solutionHashCode).toArray();
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/travelingtournament/domain/TravelingTournament.java]: "Notice: we don't use hashCode()" - why not? 
Ow... On Match.java
```
    /**
     * The normal methods {@link #equals(Object)} and {@link #hashCode()} cannot be used
     * because the rule engine already requires them (for performance in their original state).
     * @see #solutionHashCode()
     */
    public boolean solutionEquals(Object o) {
```
Ok, that didn't age well... (the code was written in 2007 or even 2006, because it's part of the first commit that came from SourceForge)
That javadocs is probably not true any more. Drools doesn't use Java equality since Drools 4 or so.
I've created a jira: https://issues.redhat.com/browse/PLANNER-1746

**Comment by triceo** [comment]
@ge0ffrey Please re-review, the changes were significant.
Also, if merging, please squash.

File: optaplanner-core/src/main/java/org/optaplanner/core/api/score/buildin/simple/SimpleScore.java [hunk 18]
```diff
@@ -169,15 +170,11 @@ public boolean equals(Object o) {
...
-                + initScore) * 37
-                + score;
+        return Objects.hash(initScore, score);
     }
 
     @Override
     public int compareTo(SimpleScore other) {
-        // A direct implementation (instead of CompareToBuilder) to avoid dependencies
         if (initScore != other.getInitScore()) {
             return initScore < other.getInitScore() ? -1 : 1;
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/api/score/buildin/simple/SimpleScore.java]: Nitpick: we can use return Integer.compare(initScore, other.getInitScore()); here too

File: optaplanner-examples/src/main/java/org/optaplanner/examples/travelingtournament/domain/Match.java [hunk 19]
```diff
@@ -69,17 +69,15 @@ public void setDay(Day day) {
...
-                    .append(homeTeam, other.homeTeam)
-                    .append(awayTeam, other.awayTeam)
-                    .append(day, other.day)
-                    .isEquals();
-        } else {
+        }
+        if (o == null || getClass() != o.getClass()) {
             return false;
         }
+        final Match match = (Match) o;
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/travelingtournament/domain/Match.java]: Nitpick: variable name other?

File: optaplanner-examples/src/main/java/org/optaplanner/examples/cheaptime/solver/drools/IdleCost.java [hunk 20]
```diff
@@ -58,25 +57,19 @@ public long getCost() {
...
-                    .append(machine, other.machine)
-                    .append(activePeriodAfterIdle, other.activePeriodAfterIdle)
-                    .append(cost, other.cost)
-                    .isEquals();
-        } else {
+        }
+        if (o == null || getClass() != o.getClass()) {
             return false;
         }
+        final IdleCost idleCost = (IdleCost) o;
```
Comment (ge0ffrey) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/cheaptime/solver/drools/IdleCost.java]: nitpick: other variable name
